src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs

```diff
@@ -1103,6 +1908,28 @@ private static bool GreaterThan(ref TKey left, ref TKey right)
     /// <summary>Helper methods for use in array/span sorting routines.</summary>
     internal static class SortUtils
     {
+        // Partitions below this size are sorted using insertion sort.
+        internal const int InsertionSortThreshold = 24;
+        // Partitions above this size use Tukey's ninther to select the pivot.
+        internal const int NintherThreshold = 128;
+        // When we detect an already sorted partition, attempt an insertion sort that allows this
+        // amount of element moves before giving up.
+        internal const int PartialInsertionSortLimit = 8;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static ref T UnguardedAccess<T>(Span<T> span, int index)
+        {
+            // return ref Unsafe.Add(ref MemoryMarshal.GetReference(span), index);
+            // TODO: investigate where can we use the above line with confidence instead of this.
+            return ref span[index];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static Span<T> UnguardedSlice<T>(Span<T> span, int begin, int end)
+        {
+            return MemoryMarshal.CreateSpan(ref UnguardedAccess(span, begin), end - begin);
+        }
```

Comment:
Using unsafe code throughout a complex algorithm like sort is a major liability. The existing sort is intentionally using safe bounds checked access, except a few carefully audited places.